### ISSUE TITLE
Implementation of Method I and J in Method D

### ISSUE THREAD

**Issue by JetteSchumann** [JuPedSim/jpscore#792]
Since the basic functionality of Method D, I and J (and the code) are the same, we should consider  an integration of Method I and J into Method D.
The basic function of Method D should remain the same.
By defining additional input parameters Method I and J could be realized in Method D.
Options could be:
- `use_blind_points = "True"` to enable the calculation of vornoi cells for less than 3 people as in Method I
- `calc_IFD_only = "True"` to enable the calculation as provided in Method I
- `use_alternative_v_calc = "True"` to use velocity calculation which is implemented in Method J
They should be `False` by default.

Any suggestions for input parameters?

**Comment by chraibi** [comment]
Maybe `use_blind_points=True` for all?

**Comment by JetteSchumann** [comment]
@chraibi you mean using blind points by default for all methods?

**Comment by chraibi** [comment]
Yes, exactly. Even with Method D, you sometimes want to measure the FD of a, say 1D experiment. In low densities, you don't have always 3 people in the measurement area at once.

**Comment by gjaeger** [comment]
At first: I like Jette's suggestion. 👍 

For a proposal for the combination of methods I would like to start with a small study.

Starting from the _Methods_ test suite in _jpsreport/Utest_, I have performed the following test. I have taken into account the result of issue #772 that the profiles are created for the entire geometry. 

- geometry with three subrooms ([geo_Methods_subrooms.xml.zip](https://github.com/JuPedSim/jpscore/files/4680823/geo_Methods_subrooms.xml.zip))
- inifile for geometry with thee subrooms ([ini_Methods_subrooms.xml.zip](https://github.com/JuPedSim/jpscore/files/4680841/ini_Methods_subrooms.xml.zip))

My expectations at the start:
- Profiles over the entire geometry across all subrooms.
- Individual data (IFD) for Method I over the entire geometry across all subrooms.

The considered measuring areas:
```xml
  <measurement_areas unit="m">

    <area_B id="1" type="BoundingBox" zPos="None">
    </area_B>

    <area_B id="5" type="BoundingBox" zPos="None">
      <vertex x="20" y="12" />
      <vertex x="24" y="12" />
      <vertex x="24" y="8" />
      <vertex x="20" y="8" />
    </area_B>

    <area_B id="9" type="BoundingBox" zPos="None">
      <vertex x="16" y="12" />
      <vertex x="20" y="12" />
      <vertex x="20" y="8" />
      <vertex x="16" y="8" />
    </area_B>
    
    <area_B id="10" type="BoundingBox" zPos="None">
      <vertex x="28" y="12" />
      <vertex x="32" y="12" />
      <vertex x="32" y="8" />
      <vertex x="28" y="8" />
    </area_B>

  </measurement_areas>
```

An excerpt of the input file:

```xml
  <method_D enabled="true">
    <measurement_area id="1" start_frame="None" stop_frame="None" get_individual_FD="true"/>
    <measurement_area id="5" start_frame="None" stop_frame="None" get_individual_FD="true"/>
    <measurement_area id="8" start_frame="None" stop_frame="None" get_individual_FD="true"/>
    <measurement_area id="9" start_frame="None" stop_frame="None" get_individual_FD="true"/>
    <measurement_area id="10" start_frame="None" stop_frame="None" get_individual_FD="true"/>
    <one_dimensional enabled="false"/>
    <cut_by_circle enabled="true" radius="1.0" edges="10"/>
    <profiles enabled="true" grid_size_x="0.20" grid_size_y="0.20"/>
  </method_D>

  <method_I enabled="true">
    <measurement_area id="1" start_frame="None" stop_frame="None" get_individual_FD="true"/>
    <cut_by_circle enabled="true" radius="1.0" edges="10"/>
	</method_I>

  <method_J enabled="true">
    <measurement_area id="1" start_frame="None" stop_frame="None" get_individual_FD="true"/>
    <measurement_area id="5" start_frame="None" stop_frame="None" get_individual_FD="true"/>
    <one_dimensional enabled="false"/>
    <cut_by_circle enabled="true" radius="1.0" edges="10"/>
    <profiles enabled="true" grid_size_x="0.20" grid_size_y="0.20"/>
  </method_J>
```

In the following comments I give the results.

**Comment by gjaeger** [comment]
# profiles - density

The profiles are independent of the measurement area (see issue #772). They extend over the entire geometry across all subrooms.

## measurement area 1
![profile_id_1_density](https://user-images.githubusercontent.com/7759675/82879508-5371a680-9f3d-11ea-9179-fae07449de39.png)

## measurement area 5
![profile_id_5_density](https://user-images.githubusercontent.com/7759675/82879568-5ff5ff00-9f3d-11ea-87f6-256766bde2fb.png)

## measurement area 9
![profile_id_9_density](https://user-images.githubusercontent.com/7759675/82879585-67b5a380-9f3d-11ea-962f-9c9cb54a7419.png)

# ⚠️ Warning ⚠️ 
If measuring areas are used for different methods, the profile is created anew for each method. Existing files are overwritten.

**Comment by gjaeger** [comment]
# profiles - velocity

The methods [D](https://www.jupedsim.org/jpsreport_method_D.html) and [J](https://www.jupedsim.org/jpsreport_method_J.html) differ in the calculation of the mean velocity. 

In the current version, the profiles for the different methods are overwritten. This should be noted.  The profiles for methods D and J are as follows:

## measurement area 9 - method D
![profile_v_id_9](https://user-images.githubusercontent.com/7759675/82895402-75c2ee80-9f54-11ea-985a-4b93a1efc39f.png)

## measurement area 5 - method J
![profile_v_id_5](https://user-images.githubusercontent.com/7759675/82895407-7a87a280-9f54-11ea-8d65-36e1f00c3764.png)

## summary 
There are differences.
⚠️ We should take this knowledge into account in our further work.

**Comment by gjaeger** [comment]
# trajectories based on individual data

## method I
[Method I](http://www.jupedsim.org/jpsreport_inifile.html#method-i) can be used for time-series analysis of individual data. Across the boundaries of the subrooms, the individual data is determined and stored in the file `IFD_I`.

![traj_I_1](https://user-images.githubusercontent.com/7759675/82897801-7cebfb80-9f58-11ea-8092-fa7862ea1214.png)

I had expected this result.

## method D
The individual data for agents that have an intersection (with their voronoi cell) with the measurement area are stored in the file `IFD_D`. The data is limited to the area of the measuring surface and depends on the setting.

### measurement area 5
![traj_D_5](https://user-images.githubusercontent.com/7759675/82898313-54183600-9f59-11ea-96d4-f8e5855cb2cf.png)

### measurement area 9
![traj_D_9](https://user-images.githubusercontent.com/7759675/82898321-59758080-9f59-11ea-8352-f5db8577b225.png)

### measurement area 10
![traj_D_10](https://user-images.githubusercontent.com/7759675/82898326-5e3a3480-9f59-11ea-8f6c-e9b038e31bc6.png)

I had expected these result. We should take these differences into account.

**Comment by gjaeger** [comment]
When comparing the velocity profiles of methods D and J, you can think to recognize small differences. The detailed view for one frame (e.g. frame 21-22) illustrates the differences.

## profile based on method J - measurement area 5
![vel_J_id_5_fr_21-22](https://user-images.githubusercontent.com/7759675/82978531-1b25a300-9fe5-11ea-8f0c-d84fe5656398.png)

## profile based on method D - measurement area 9
![vel_D_id_9_fr_21-22](https://user-images.githubusercontent.com/7759675/82978606-43150680-9fe5-11ea-8aac-215a59e16e59.png)

The velocity profile based on method D shows blurred edges. This becomes clear in a detail view (first: method J - measurement area 5; second: method D - measurement area 9):

![vel_J_id_5_fr_21-22_detail](https://user-images.githubusercontent.com/7759675/82978749-9b4c0880-9fe5-11ea-8dd4-07f8d043e053.png)

![vel_D_id_9_fr_21-22_detail](https://user-images.githubusercontent.com/7759675/82978825-c3d40280-9fe5-11ea-8431-5c34c4ba9079.png)

The difference is due to the method being called in the `GetProfiles` function (method D: [line 364](https://github.com/JuPedSim/jpscore/blob/master/jpsreport/methods/Method_D.cpp#L424); method J: [line 368](https://github.com/JuPedSim/jpscore/blob/master/jpsreport/methods/Method_J.cpp#L368)). When calculating the profiles, the grid cells are used as measurement areas and the different velocity functions `GetVoronoiDensityVelocity` (method D: [line 328](https://github.com/JuPedSim/jpscore/blob/master/jpsreport/methods/Method_D.cpp#L328); method J: [line 326](https://github.com/JuPedSim/jpscore/blob/master/jpsreport/methods/Method_J.cpp#L326)) are called for the methods. @JetteSchumann sees great potential for optimisation in these functions.

By the way, the profiles for the density are the same, since the methods of calculation do not differ.

I ask myself the basic questions: 
- Is the calculation of the profiles correct?
- Is there a faster calculation algorithm?

During the implementation we have to document the calculation method.

How can we define the requirements for an implementation and create a work plan?

**Comment by gjaeger** [comment]
To my question "_Is the calculation of the profiles correct?_" I was asked why I ask this. I would like to clarify my question:

With the methods for a measurement area we calculate the mean density and velocity for each grid cell. Afterwards we use an interpolation method when plotting with Matplotlib. Do we need this accuracy when we interpolate at the end? Is there perhaps a faster algorithm if we use the individual data (`IFD_I`)? The calculation of individual data with jpsreport is fast. Maybe a python script for the profiles is better? Does anyone know a good solution?

**Comment by chraibi** [comment]
> Afterwards we use an interpolation method when plotting with Matplotlib

which interpolation?

> Maybe a python script for the profiles is better?

depends on your definition of "profiles". Could you give more details about what you suggest?

**Comment by gjaeger** [comment]
> > Afterwards we use an interpolation method when plotting with Matplotlib
> 
> which interpolation?

see [/jpsreport/scripts/_Plot_profiles.py, line 81](https://github.com/JuPedSim/jpscore/blob/master/jpsreport/scripts/_Plot_profiles.py#L81): `interpolation = 'nearest'` 
```python
   im = plt.imshow(density, cmap=cm.jet,
                    interpolation='nearest', origin='lower',
                    vmin=0, vmax=np.amax(density),
                    extent=[geominX, geomaxX, geominY, geomaxY])
```
 
> > Maybe a python script for the profiles is better?
> 
> depends on your definition of "profiles". Could you give more details about what you suggest?

In the following I refer to the individual data (`IFD_I`) and the default setting of cell size dx=dy=0.2 Assuming that an agent can occupy only one cell, one could iterate over the agents for each frame and assign the individual velocity or density to the cell. When creating the profile, the intermediate values between the central cells could be set via the interpolation setting of `imshow`. One would have to look at the possible difference. 

This is a first thought.

**Comment by chraibi** [comment]
If I understand correctly you propose an alternative script, not only to plot the profiles but to calculate them in a different way. If the methodology is correct and the script is fast enough, we could remove the calculation in jpsreport and outsource it to the script. 

Wanna give it a try?

**Comment by gjaeger** [comment]
Using the example of the demo "[_bottleneck_](https://github.com/JuPedSim/jpscore/tree/master/jpsreport/demos/bottleneck)", I made an analysis with the profile files and the `IFD_I` file. The results can be found in the [notebook (ipynb) on gists](https://gist.github.com/gjaeger/23d13448eea498ee118c9d5cec891043) for issue #576.

**Comment by gjaeger** [comment]
> - `velocity_calculation = "XYZ"` to use velocity calculation which is implemented in Method J
>
> They should be False by default.
> Any suggestions for input parameters?

I suggest the following parameters for veleocity calculation:
- `"voronoi"` for Method D
- `"arithmetic"` for Method J

**Comment by gjaeger** [comment]
in addition to my last comment ... I propose to change the documentation of the velocity calculation based on the formulas describing the previous methods C and J as follows:

[method C ](https://www.jupedsim.org/jpsreport_method_C.html)
<img width="305" alt="method_c_vel-calc" src="https://user-images.githubusercontent.com/7759675/87524071-04680800-c688-11ea-98f8-ed65b62b2b9c.png">
 
[method J](https://www.jupedsim.org/jpsreport_method_J.html)
<img width="493" alt="method_J_vel-calc" src="https://user-images.githubusercontent.com/7759675/87524222-2c576b80-c688-11ea-854a-6b60e49622f9.png">

new description:
<img width="452" alt="method_vel-calc_new" src="https://user-images.githubusercontent.com/7759675/87524445-780a1500-c688-11ea-9184-2f5794bb54ad.png">

```tex
$\langle v \rangle_{A}=\frac{1}{N_{(x,y) \in A}}\sum_{i=1}^{N_{(x,y) \in A}}{v_i(t)}$
```

**Comment by gjaeger** [comment]
After dicussion with Jette, our proposal for the realisation of issue #808, #809 and #810:

```xml
<measurement_method_voronoi enabled="true">
   <measurement_area id="1" start_frame="None" stop_frame="None" local_individual="true"/>
   <measurement_area id="2" start_frame="None" stop_frame="None" local_individual="true"/>
   <cut_by_circle enabled="false" radius="1.0" edges="10"/>
   <global_individual enabled="true" zPos="None"/>
   <profiles enabled="false" zPos="None" grid_size_x="0.20" grid_size_y="0.20"/>
 </measurement_method_voronoi>
```

This includes default settings:
```xml
   <vel_calculation="Voronoi"/>
   <blind_points enabled="true"/>
```

Current [alternatives](https://github.com/JuPedSim/jpscore/issues/792#issuecomment-658618438):
```xml
   <vel_calculation="Arithmetic"/>
   <blind_points enabled="false"/>
```

**Comment by gjaeger** [comment]
As a reminder, when restructuring, we should also review the need for the `jpsreport/Utest` folder.

**Comment by gjaeger** [comment]
> After dicussion with Jette, our proposal for the realisation of issue #808, #809 and #810:
> 
> ```
> <measurement_method_voronoi enabled="true">
>    <measurement_area id="1" start_frame="None" stop_frame="None" local_individual="true"/>
>    <measurement_area id="2" start_frame="None" stop_frame="None" local_individual="true"/>
>    <cut_by_circle enabled="false" radius="1.0" edges="10"/>
>    <global_individual enabled="true" zPos="None"/>
>    <profiles enabled="false" zPos="None" grid_size_x="0.20" grid_size_y="0.20"/>
>  </measurement_method_voronoi>
> ```
> 
> This includes default settings:
> 
> ```
>    <vel_calculation="Voronoi"/>
>    <blind_points enabled="true"/>
> ```
> 
> Current [alternatives](https://github.com/JuPedSim/jpscore/issues/792#issuecomment-658618438):
> 
> ```
>    <vel_calculation="Arithmetic"/>
>    <blind_points enabled="false"/>
> ```

The following changes were made and are to be considered in the documentation (PR #840):
- old: `local_individual="true"` to **new:** `local_IFD="true"`
- old: `<global_individual enabled="true" zPos="None"/>`to **new:** `<global_IFD enabled="true"/>`
- old: `<blind_points enabled="false"/>` to **new:** `<use_blind_points enabled="false"/>`

The parameter `zPos` **must** be checked after refactoring.

**Comment by chraibi** [comment]
thanks Gregor and Jette. 
That happens when `global_IFD enabled="true"` and  `local_IFD enabled="true"`?

**Comment by JetteSchumann** [comment]
@chraibi data are calculated twice. Once for the whole geometry (global) and once for the specified measurement area (local)

**Comment by gjaeger** [comment]
> thanks Gregor and Jette.
> That happens when `global_IFD enabled="true"` and `local_IFD enabled="true"`?

With `global_IFD enabled="true"` the data for an IFD_file for the entire geometry is created. With `local_IFD enabled="true"`, the corresponding measurement area an IFD file related to the measurement area is created with the information about the intersection polygon.

**Comment by chraibi** [comment]
Okay, thanks.
 Is there any scenario where the following setup does make sense?

`<measurement_area id="1" start_frame="None" stop_frame="None" local_individual="false"/>`

**Comment by JetteSchumann** [comment]
yes. that makes sense if you are only interested in voronoi density and velocity but not the individual data
